File: track_b_systems/python_core/week2/day11/storage_system.py
import json
import os

#======= CONFIG LAYER =======#

DATA_FILE="user_data.json"

#======= DATA MODEL =======#

class User:
	def __init__(self,username,pin,balance=1000,daily_withdrawn=0):
		self.username=username
		self.pin=pin
		self.balance=balance
		self.daily_withdrawn=daily_withdrawn

	def to_dict(self):
		return {
			"username":self.username,
			"pin":self.pin,
			"balance":self.balance,
			"daily_withdrawn":self.daily_withdrawn
		}
	
	@staticmethod
	def from_dict(data):
		return User(
			username=data["username"],
			pin=data["pin"],
			balance=data["balance"],
			daily_withdrawn=data["daily_withdrawn"]
		)
	
#======= STORAGE LAYER =======#

class StorageManager:
	@staticmethod
	def save_user(user: User):
		with open (DATA_FILE,"w") as f:
			json.dump(user.to_dict(),f,indent=4)

	@staticmethod
	def load_user():
		if not os.path.exists(DATA_FILE):
			return None
		
		with open(DATA_FILE,"r") as f:
			data=json.load(f)
			return User.from_dict(data)
		
#======= AUTH LAYER =======#

def authenticate(user : User):
	for i in range(3,0,-1):
		name=input("Enter the username : ")
		pin=input("Enter pin : ")
		
		if name==user.username and pin ==user.pin:
			print("Login succesful. \n") 
			return True

		else :
			print(f"Wrong credentials. Attempts left : {i-1}. \n")

	return False
	
#======= BUSINESS LOGIC =======#

def check_balance(user):
	print("Balance : ",user.balance)

def deposit(user):
	amt=int(input("Enter amount : "))
	if amt>0:
		user.balance+=amt
		print("Deposited : ",amt)

	else :
		print("Invalid amount.")

def withdraw(user):
	amt=int(input("Enter anount : "))
	if amt<=0:
		print("Invalid amount.")
	
	elif amt > user.balance:
		print("Insuffecient Balance.")

	else :
		user.balance-=amt
		user.daily_withdrawn+=amt
		print("Withdrawn : ",amt)

#======== CONTROLLER =======#

def system_controller(user):
	system_on=True
	while system_on:
		print("\n--- SYSTEM PANEL ---\n")
		print("1. Balance")
		print("2. Deposit")
		print("3. Withdraw")
		print("4. Exit")

		choice=int(input("Choose : "))

		if choice==1:
			check_balance(user)

		elif choice==2:
			deposit(user)

		elif choice==3:
			withdraw(user)

		elif choice==4:
			print("Saving system state...")
			StorageManager.save_user(user)		#Persistance on shutdown
			system_on=False

		else:
			print("Invalid Choice.")
	
#======== SYSTEM LIFECYCLE =======#

def system_boot():
	user=StorageManager.load_user()
	if user is None:
		print("No data found. Creating new user.")

		user=User(username="Raghu",pin="sarghi919")
		StorageManager.save_user(user)
	return user

def system_run(user):
	if authenticate(user):
		system_controller(user)

	else :
		print("System Locked.")
	
def system_shutdown():
	print("System shutdown complete.")

#======== MAIN =======#

def main():
	user=system_boot()		#LOAD STATE
	system_run(user)		#RUN SYSTEM
	system_shutdown()		#SHUTDOWN

if __name__=="__main__":
	main()

File: track_a_engineering/cpp_dsa/week2/day11/storage_system.cpp
#include <iostream>
#include <string>
using namespace std;

//================ CONFIG LAYER =================//

const int MAX_LIMIT = 10000;
const int DAILY_LIMIT = 25000;

//================ DATA MODEL ===================//

struct User {
    string username;
    string pin;
    int balance;
    int daily_withdrawn;
};

//================ AUTH LAYER ===================//

bool authenticate(User &u) {
    string uname, pass;

    cout << "Enter username: ";
    cin >> uname;

    for(int i = 3; i > 0; i--) {
        cout << "Enter PIN: ";
        cin >> pass;

        if(uname == u.username && pass == u.pin) {
            cout << "\nAuthentication successful!\n";
            return true;
        } else {
            cout << "Wrong credentials. Attempts left: " << i-1 << "\n";
        }
    }
    return false;
}

//================ UI LAYER =====================//

void show_menu() {
    cout << "\n====== SYSTEM PANEL ======\n";
    cout << "1. Check Balance\n";
    cout << "2. Deposit\n";
    cout << "3. Withdraw\n";
    cout << "4. Exit\n";
}

//================ BUSINESS LOGIC ===============//

void check_balance(User &u) {
    cout << "Balance: " << u.balance << endl;
}

void deposit(User &u, int amount) {
    if(amount > 0) {
        u.balance += amount;
        cout << "Deposited: " << amount << endl;
    } else {
        cout << "Invalid amount\n";
    }
}

bool can_withdraw(User &u, int amount) {
    if(amount <= 0) return false;
    if(amount > u.balance) return false;
    if(amount > MAX_LIMIT) return false;
    if(u.daily_withdrawn + amount > DAILY_LIMIT) return false;
    return true;
}

void withdraw(User &u, int amount) {
    u.balance -= amount;
    u.daily_withdrawn += amount;
    cout << "Withdrawn: " << amount << endl;
}

//================ CONTROLLER ===================//

void controller(User &u) {
    bool system_on = true;

    while(system_on) {
        show_menu();
        int choice;
        cout << "Choose option: ";
        cin >> choice;

        if(choice == 1) {
            check_balance(u);
        }
        else if(choice == 2) {
            int amt;
            cout << "Enter deposit amount: ";
            cin >> amt;
            deposit(u, amt);
        }
        else if(choice == 3) {
            int amt;
            cout << "Enter withdraw amount: ";
            cin >> amt;

            if(can_withdraw(u, amt)) {
                withdraw(u, amt);
            } else {
                cout << "Withdrawal denied (rule violation)\n";
            }
        }
        else if(choice == 4) {
            cout << "System shutting down...\n";
            system_on = false;
        }
        else {
            cout << "Invalid option\n";
        }
    }
}

//================ MAIN (BOOT) ==================//

int main() {

    User user;
    user.username = "Raghu";
    user.pin = "sarghi919";
    user.balance = 10000;
    user.daily_withdrawn = 0;

    if(authenticate(user)) {
        controller(user);
    } else {
        cout << "System locked.\n";
    }

    return 0;
}